clippy_lints/src/slice_as_bytes.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet_with_applicability, ty::is_type_lang_item};
+use rustc_errors::Applicability::MachineApplicable;
+use rustc_hir::{Expr, ExprKind, LangItem, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for string slices immediantly followed by `as_bytes`.
+    /// ### Why is this bad?
+    /// This involves doing an unnecessary UTF-8 alignment check which is less efficient, and can cause a panic.
+    /// ### Example
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// s[1..5].as_bytes();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// &s.as_bytes()[1..5];
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub SLICE_AS_BYTES,
+    nursery,
```

Comment:
It should probably not be `nursery`. `nursery` is left for lints that need further discussion / improving before going public. It should be `pedantic` imo

Reply:
Seems like it could also be `perf`, maybe? https://godbolt.org/z/qd8r8x6hW shows that doing this looks much better perf wise (1 check vs 4 checks in the happy path). Though this is a micro optimization, and I'm not sure if this is enough for the perf category, since most of the perf lints have much bigger impact (something like allocating when it's not necessary). It does seem easy enough of a fix and does not really hurt readability so maybe its fine?